File: src/II014.cpp
#include <iostream>
#include <vector>
using namespace std;

// 双指针法，没想到，感觉太巧妙了！
class Solution {
public:
  bool checkInclusion(string s1, string s2) {
    int n = s1.length();
    int m = s2.length();
    if (n > m)
      return false;
    int cnt[26] = {0};
    for (char c : s1) {
      --cnt[c - 'a'];
    }
    int l = 0;
    for (int r = 0; r < m; ++r) {
      int x = s2[r] - 'a';
      ++cnt[x];
      while (cnt[x] > 0) {
        --cnt[s2[l] - 'a'];
        ++l;
      }
      if (r - l + 1 == n)
        return true;
    }
    return false;
  }
};

File: src/II014.rs
// 自己想的滑动窗口，和题解几乎一模一样
struct Solution;

impl Solution {
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let (n, m) = (s1.len(), s2.len());
        if n > m {
            return false;
        }
        let (s1, s2) = (s1.as_bytes(), s2.as_bytes());
        let mut cnt = [0; 26];
        for i in 0..n {
            cnt[(s1[i] - b'a') as usize] += 1;
            cnt[(s2[i] - b'a') as usize] -= 1;
        }
        let mut diff = 0;
        cnt.iter().for_each(|x| {
            if *x != 0 {
                diff += 1
            }
        });
        if diff == 0 {
            return true;
        }
        for i in n..m {
            let x = (s2[i] - b'a') as usize;
            if cnt[x] == 0 {
                diff += 1;
            }
            cnt[x] -= 1;
            if cnt[x] == 0 {
                diff -= 1;
            }
            let x = (s2[i - n] - b'a') as usize;
            if cnt[x] == 0 {
                diff += 1;
            }
            cnt[x] += 1;
            if cnt[x] == 0 {
                diff -= 1;
            }
            if diff == 0 {
                return true;
            }
        }

        false
    }
}
